File: backend/internal/handlers/recipes.go
package handlers

import (
	"database/sql"
	"net/http"

	"github.com/gin-gonic/gin"
	"github.com/google/uuid"
	"pos-public/internal/models"
)

type RecipeHandler struct {
	DB *sql.DB
}

func NewRecipeHandler(db *sql.DB) *RecipeHandler {
	return &RecipeHandler{DB: db}
}

// GetProductIngredients retrieves all ingredients for a specific product
func (h *RecipeHandler) GetProductIngredients(c *gin.Context) {
	productIDStr := c.Param("id")
	productID, err := uuid.Parse(productIDStr)
	if err != nil {
		errMsg := err.Error()
		c.JSON(http.StatusBadRequest, models.APIResponse{
			Success: false,
			Message: "Invalid product ID",
			Error:   &errMsg,
		})
		return
	}

	query := `
		SELECT
			pi.id,
			pi.product_id,
			pi.ingredient_id,
			pi.quantity_required,
			i.name as ingredient_name,
			i.current_stock,
			i.unit as ingredient_unit
		FROM product_ingredients pi
		JOIN ingredients i ON pi.ingredient_id = i.id
		WHERE pi.product_id = $1
		ORDER BY i.name
	`

	rows, err := h.DB.Query(query, productID)
	if err != nil {
		errMsg := err.Error()
		c.JSON(http.StatusInternalServerError, models.APIResponse{
			Success: false,
			Message: "Failed to retrieve product ingredients",
			Error:   &errMsg,
		})
		return
	}
	defer rows.Close()

	var ingredients []models.ProductIngredient
	for rows.Next() {
		var pi models.ProductIngredient
		err := rows.Scan(
			&pi.ID,
			&pi.ProductID,
			&pi.IngredientID,
			&pi.QuantityRequired,
			&pi.IngredientName,
			&pi.CurrentStock,
			&pi.IngredientUnit,
		)
		if err != nil {
			errMsg := err.Error()
			c.JSON(http.StatusInternalServerError, models.APIResponse{
				Success: false,
				Message: "Failed to scan ingredient data",
				Error:   &errMsg,
			})
			return
		}
		ingredients = append(ingredients, pi)
	}

	c.JSON(http.StatusOK, models.APIResponse{
		Success: true,
		Message: "Product ingredients retrieved successfully",
		Data:    ingredients,
	})
}

// AddProductIngredient adds an ingredient to a product recipe
func (h *RecipeHandler) AddProductIngredient(c *gin.Context) {
	productIDStr := c.Param("id")
	productID, err := uuid.Parse(productIDStr)
	if err != nil {
		errMsg := err.Error()
		c.JSON(http.StatusBadRequest, models.APIResponse{
			Success: false,
			Message: "Invalid product ID",
			Error:   &errMsg,
		})
		return
	}

	var req models.AddRecipeIngredientRequest
	if err := c.ShouldBindJSON(&req); err != nil {
		errMsg := err.Error()
		c.JSON(http.StatusBadRequest, models.APIResponse{
			Success: false,
			Message: "Invalid request body",
			Error:   &errMsg,
		})
		return
	}

	// Verify product exists
	var productExists bool
	err = h.DB.QueryRow("SELECT EXISTS(SELECT 1 FROM products WHERE id = $1)", productID).Scan(&productExists)
	if err != nil || !productExists {
		c.JSON(http.StatusNotFound, models.APIResponse{
			Success: false,
			Message: "Product not found",
		})
		return
	}

	// Verify ingredient exists
	var ingredientExists bool
	err = h.DB.QueryRow("SELECT EXISTS(SELECT 1 FROM ingredients WHERE id = $1)", req.IngredientID).Scan(&ingredientExists)
	if err != nil || !ingredientExists {
		c.JSON(http.StatusNotFound, models.APIResponse{
			Success: false,
			Message: "Ingredient not found",
		})
		return
	}

	// Check if ingredient already exists in recipe
	var exists bool
	err = h.DB.QueryRow(
		"SELECT EXISTS(SELECT 1 FROM product_ingredients WHERE product_id = $1 AND ingredient_id = $2)",
		productID, req.IngredientID,
	).Scan(&exists)
	if err != nil {
		errMsg := err.Error()
		c.JSON(http.StatusInternalServerError, models.APIResponse{
			Success: false,
			Message: "Failed to check ingredient existence",
			Error:   &errMsg,
		})
		return
	}
	if exists {
		c.JSON(http.StatusConflict, models.APIResponse{
			Success: false,
			Message: "Ingredient already exists in this product recipe",
		})
		return
	}

	// Insert product ingredient
	var newID uuid.UUID
	query := `
		INSERT INTO product_ingredients (product_id, ingredient_id, quantity_required)
		VALUES ($1, $2, $3)
		RETURNING id
	`
	err = h.DB.QueryRow(query, productID, req.IngredientID, req.QuantityRequired).Scan(&newID)
	if err != nil {
		errMsg := err.Error()
		c.JSON(http.StatusInternalServerError, models.APIResponse{
			Success: false,
			Message: "Failed to add ingredient to product",
			Error:   &errMsg,
		})
		return
	}

	c.JSON(http.StatusCreated, models.APIResponse{
		Success: true,
		Message: "Ingredient added to product successfully",
		Data:    map[string]string{"id": newID.String()},
	})
}

// UpdateProductIngredient updates the quantity required for a product ingredient
func (h *RecipeHandler) UpdateProductIngredient(c *gin.Context) {
	productIDStr := c.Param("id")
	productID, err := uuid.Parse(productIDStr)
	if err != nil {
		errMsg := err.Error()
		c.JSON(http.StatusBadRequest, models.APIResponse{
			Success: false,
			Message: "Invalid product ID",
			Error:   &errMsg,
		})
		return
	}

	ingredientIDStr := c.Param("ingredient_id")
	ingredientID, err := uuid.Parse(ingredientIDStr)
	if err != nil {
		errMsg := err.Error()
		c.JSON(http.StatusBadRequest, models.APIResponse{
			Success: false,
			Message: "Invalid ingredient ID",
			Error:   &errMsg,
		})
		return
	}

	var req models.UpdateRecipeIngredientRequest
	if err := c.ShouldBindJSON(&req); err != nil {
		errMsg := err.Error()
		c.JSON(http.StatusBadRequest, models.APIResponse{
			Success: false,
			Message: "Invalid request body",
			Error:   &errMsg,
		})
		return
	}

	query := `
		UPDATE product_ingredients
		SET quantity_required = $1, updated_at = CURRENT_TIMESTAMP
		WHERE product_id = $2 AND ingredient_id = $3
	`
	result, err := h.DB.Exec(query, req.QuantityRequired, productID, ingredientID)
	if err != nil {
		errMsg := err.Error()
		c.JSON(http.StatusInternalServerError, models.APIResponse{
			Success: false,
			Message: "Failed to update product ingredient",
			Error:   &errMsg,
		})
		return
	}

	rowsAffected, _ := result.RowsAffected()
	if rowsAffected == 0 {
		c.JSON(http.StatusNotFound, models.APIResponse{
			Success: false,
			Message: "Product ingredient not found",
		})
		return
	}

	c.JSON(http.StatusOK, models.APIResponse{
		Success: true,
		Message: "Product ingredient updated successfully",
	})
}

// DeleteProductIngredient removes an ingredient from a product recipe
func (h *RecipeHandler) DeleteProductIngredient(c *gin.Context) {
	productIDStr := c.Param("id")
	productID, err := uuid.Parse(productIDStr)
	if err != nil {
		errMsg := err.Error()
		c.JSON(http.StatusBadRequest, models.APIResponse{
			Success: false,
			Message: "Invalid product ID",
			Error:   &errMsg,
		})
		return
	}

	ingredientIDStr := c.Param("ingredient_id")
	ingredientID, err := uuid.Parse(ingredientIDStr)
	if err != nil {
		errMsg := err.Error()
		c.JSON(http.StatusBadRequest, models.APIResponse{
			Success: false,
			Message: "Invalid ingredient ID",
			Error:   &errMsg,
		})
		return
	}

	query := "DELETE FROM product_ingredients WHERE product_id = $1 AND ingredient_id = $2"
	result, err := h.DB.Exec(query, productID, ingredientID)
	if err != nil {
		errMsg := err.Error()
		c.JSON(http.StatusInternalServerError, models.APIResponse{
			Success: false,
			Message: "Failed to delete product ingredient",
			Error:   &errMsg,
		})
		return
	}

	rowsAffected, _ := result.RowsAffected()
	if rowsAffected == 0 {
		c.JSON(http.StatusNotFound, models.APIResponse{
			Success: false,
			Message: "Product ingredient not found",
		})
		return
	}

	c.JSON(http.StatusOK, models.APIResponse{
		Success: true,
		Message: "Ingredient removed from product successfully",
	})
}

File: backend/src/handlers/recipes.ts
import type { Context } from 'hono';
import { sql } from 'drizzle-orm';
import { db } from '../db/connection.js';
import { successResponse, errorResponse } from '../lib/response.js';

// ── GetProductIngredients ──────────────────────────────────────────────────────

export async function getProductIngredients(c: Context) {
  const productId = c.req.param('id');

  try {
    const rows = await db.execute<{
      id: string;
      product_id: string;
      ingredient_id: string;
      quantity_required: string;
      ingredient_name: string;
      current_stock: string;
      ingredient_unit: string;
    }>(sql`
      SELECT
        pi.id,
        pi.product_id,
        pi.ingredient_id,
        pi.quantity_required,
        i.name as ingredient_name,
        i.current_stock,
        i.unit as ingredient_unit
      FROM product_ingredients pi
      JOIN ingredients i ON pi.ingredient_id = i.id
      WHERE pi.product_id = ${productId}
      ORDER BY i.name
    `);

    const ingredients = rows.rows.map((row) => ({
      id: row.id,
      product_id: row.product_id,
      ingredient_id: row.ingredient_id,
      quantity_required: Number(row.quantity_required),
      ingredient_name: row.ingredient_name,
      current_stock: Number(row.current_stock),
      ingredient_unit: row.ingredient_unit,
    }));

    return successResponse(c, 'Product ingredients retrieved successfully', ingredients);
  } catch (err) {
    return errorResponse(c, 'Failed to retrieve product ingredients', (err as Error).message);
  }
}

// ── AddProductIngredient ──────────────────────────────────────────────────────

export async function addProductIngredient(c: Context) {
  const productId = c.req.param('id');

  let body: {
    ingredient_id: string;
    quantity_required: number;
  };

  try {
    body = await c.req.json();
  } catch {
    return errorResponse(c, 'Invalid request body', 'invalid_json', 400);
  }

  if (!body.ingredient_id) {
    return errorResponse(c, 'ingredient_id is required', 'missing_ingredient_id', 400);
  }
  if (!body.quantity_required || body.quantity_required <= 0) {
    return errorResponse(c, 'quantity_required must be greater than 0', 'invalid_quantity', 400);
  }

  try {
    // Verify product exists
    const productRes = await db.execute<{ exists: boolean }>(sql`
      SELECT EXISTS(SELECT 1 FROM products WHERE id = ${productId}) as exists
    `);
    if (!productRes.rows[0]?.exists) {
      return errorResponse(c, 'Product not found', 'product_not_found', 404);
    }

    // Verify ingredient exists
    const ingredientRes = await db.execute<{ exists: boolean }>(sql`
      SELECT EXISTS(SELECT 1 FROM ingredients WHERE id = ${body.ingredient_id}) as exists
    `);
    if (!ingredientRes.rows[0]?.exists) {
      return errorResponse(c, 'Ingredient not found', 'ingredient_not_found', 404);
    }

    // Check if ingredient already exists in recipe
    const existsRes = await db.execute<{ exists: boolean }>(sql`
      SELECT EXISTS(
        SELECT 1 FROM product_ingredients WHERE product_id = ${productId} AND ingredient_id = ${body.ingredient_id}
      ) as exists
    `);
    if (existsRes.rows[0]?.exists) {
      return errorResponse(c, 'Ingredient already exists in this product recipe', 'duplicate_ingredient', 409);
    }

    // Insert product ingredient
    const insertRes = await db.execute<{ id: string }>(sql`
      INSERT INTO product_ingredients (product_id, ingredient_id, quantity_required)
      VALUES (${productId}, ${body.ingredient_id}, ${body.quantity_required})
      RETURNING id
    `);

    return successResponse(c, 'Ingredient added to product successfully', {
      id: insertRes.rows[0].id,
    }, 201);
  } catch (err) {
    return errorResponse(c, 'Failed to add ingredient to product', (err as Error).message);
  }
}

// ── UpdateProductIngredient ──────────────────────────────────────────────────────

export async function updateProductIngredient(c: Context) {
  const productId = c.req.param('id');
  const ingredientId = c.req.param('ingredient_id');

  let body: {
    quantity_required: number;
  };

  try {
    body = await c.req.json();
  } catch {
    return errorResponse(c, 'Invalid request body', 'invalid_json', 400);
  }

  if (!body.quantity_required || body.quantity_required <= 0) {
    return errorResponse(c, 'quantity_required must be greater than 0', 'invalid_quantity', 400);
  }

  try {
    const res = await db.execute(sql`
      UPDATE product_ingredients
      SET quantity_required = ${body.quantity_required}, updated_at = CURRENT_TIMESTAMP
      WHERE product_id = ${productId} AND ingredient_id = ${ingredientId}
    `);

    if (res.rowCount === 0) {
      return errorResponse(c, 'Product ingredient not found', 'not_found', 404);
    }

    return successResponse(c, 'Product ingredient updated successfully');
  } catch (err) {
    return errorResponse(c, 'Failed to update product ingredient', (err as Error).message);
  }
}

// ── DeleteProductIngredient ──────────────────────────────────────────────────────

export async function deleteProductIngredient(c: Context) {
  const productId = c.req.param('id');
  const ingredientId = c.req.param('ingredient_id');

  try {
    const res = await db.execute(sql`
      DELETE FROM product_ingredients WHERE product_id = ${productId} AND ingredient_id = ${ingredientId}
    `);

    if (res.rowCount === 0) {
      return errorResponse(c, 'Product ingredient not found', 'not_found', 404);
    }

    return successResponse(c, 'Ingredient removed from product successfully');
  } catch (err) {
    return errorResponse(c, 'Failed to delete product ingredient', (err as Error).message);
  }
}
